homeassistant/components/dsmr/sensor.py

```diff
@@ -67,153 +76,51 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
-        ["Voltage Swells Phase L2", obis_ref.VOLTAGE_SWELL_L2_COUNT, False],
-        ["Voltage Swells Phase L3", obis_ref.VOLTAGE_SWELL_L3_COUNT, False],
-        ["Voltage Phase L1", obis_ref.INSTANTANEOUS_VOLTAGE_L1, False],
-        ["Voltage Phase L2", obis_ref.INSTANTANEOUS_VOLTAGE_L2, False],
-        ["Voltage Phase L3", obis_ref.INSTANTANEOUS_VOLTAGE_L3, False],
-        ["Current Phase L1", obis_ref.INSTANTANEOUS_CURRENT_L1, False],
-        ["Current Phase L2", obis_ref.INSTANTANEOUS_CURRENT_L2, False],
-        ["Current Phase L3", obis_ref.INSTANTANEOUS_CURRENT_L3, False],
-    ]
-
-    if dsmr_version == "5L":
-        obis_mapping.extend(
-            [
-                [
-                    "Energy Consumption (total)",
-                    obis_ref.LUXEMBOURG_ELECTRICITY_USED_TARIFF_GLOBAL,
-                    True,
-                ],
-                [
-                    "Energy Production (total)",
-                    obis_ref.LUXEMBOURG_ELECTRICITY_DELIVERED_TARIFF_GLOBAL,
-                    True,
-                ],
-            ]
-        )
-    else:
-        obis_mapping.extend(
-            [["Energy Consumption (total)", obis_ref.ELECTRICITY_IMPORTED_TOTAL, True]]
-        )
-
-    # Generate device entities
+    dsmr_version = entry.data[CONF_DSMR_VERSION]
     devices = [
-        DSMREntity(
-            name, DEVICE_NAME_ENERGY, config[CONF_SERIAL_ID], obis, config, force_update
-        )
-        for name, obis, force_update in obis_mapping
+        DSMREntity(sensor, entry)
+        for sensor in SENSORS
+        if (sensor.dsmr_versions is None or dsmr_version in sensor.dsmr_versions)
+        and (not sensor.is_gas or CONF_SERIAL_ID_GAS in entry.data)
     ]
-
-    # Protocol version specific obis
-    if CONF_SERIAL_ID_GAS in config:
-        if dsmr_version in ("4", "5", "5L"):
-            gas_obis = obis_ref.HOURLY_GAS_METER_READING
-        elif dsmr_version in ("5B",):
-            gas_obis = obis_ref.BELGIUM_HOURLY_GAS_METER_READING
-        else:
-            gas_obis = obis_ref.GAS_METER_READING
-
-        # Add gas meter reading
-        devices += [
-            DSMREntity(
-                "Gas Consumption",
-                DEVICE_NAME_GAS,
-                config[CONF_SERIAL_ID_GAS],
-                gas_obis,
-                config,
-                True,
-            )
-        ]
-
     async_add_entities(devices)
 
     min_time_between_updates = timedelta(
-        seconds=options.get(CONF_TIME_BETWEEN_UPDATE, DEFAULT_TIME_BETWEEN_UPDATE)
+        seconds=entry.options.get(CONF_TIME_BETWEEN_UPDATE, DEFAULT_TIME_BETWEEN_UPDATE)
     )
 
     @Throttle(min_time_between_updates)
-    def update_entities_telegram(telegram):
+    def update_entities_telegram(telegram: dict[str, DSMRObject]):
         """Update entities with latest telegram and trigger state update."""
         # Make all device entities aware of new telegram
         for device in devices:
```

Comment:
This should preferably be refactored next to use our dispatch helper instead.

Reply:
Yup 👍 

Baby steps :)